experimental/Schemes/Tjurina.jl

```diff
@@ -0,0 +1,751 @@
+    return tjurina_number(f) != PosInf()
+  end
+end
+
+
+
+
+@doc raw"""
+    is_finitely_determined(X::HypersurfaceGerm, equivalence::Symbol = :contact)
+
+Return if the hypersurface germ 'X' is finitely determined with respect to ':right' or ':contact' equivalence. 
+By default computes with respect to contact equivalence.
+# Examples
+```jldoctest
+julia> R,(x,y) = QQ["x", "y"];
+
+julia> f = x^2 - y^2;
+
+julia> X = HypersurfaceGerm(AffineScheme(quo(R, ideal(R, f))[1]), [0, 0]);
+
+julia> is_finitely_determined(X)
+true
+
+julia> is_finitely_determined(X, :right)
+true
+```
+"""
+function is_finitely_determined(X::HypersurfaceGerm, equivalence::Symbol = :contact)
+  f = defining_ideal(X)[1]
+  return is_finitely_determined(f, equivalence)
+end
+
+
+
+###############################################################################
+@doc raw"""
+    determinacy_bound(f::MPolyLocRingElem, equivalence::Symbol = :contact)
+
+Compute some determinacy bound of 'f' with respect to ':right' or ':contact' equivalence.
+Return infinity if not finitely determined. 
+By default computes with respect to contact equivalence.
+This computation is based on the Milnor number respectively Tjurina number.
+# Examples
+```jldoctest
+julia> R,(x,y) = QQ["x", "y"];
+
+julia> L,_  = localization(R, complement_of_point_ideal(R, [0, 0]));
+
+julia> f = L(x^3 - y^2);
+
+julia> determinacy_bound(f)
+3
+
+julia> determinacy_bound(f, :right)
+3
+```
+"""
+function determinacy_bound(f::MPolyLocRingElem, equivalence::Symbol = :contact)
+  equivalence == :right || equivalence == :contact || error("Equivalence typ must be ':right' or ':contact'.")
+  ord_f = order_as_series(f)
+  ## if the order of f is 1, then f is right and contact equivalent to its 1-jet (smooth case)
+  ord_f != 1 || return 1
+  is_finitely_determined(f, equivalence) || return PosInf()
+  if equivalence == :right
+    if ord_f == 0     
+      ## A unit has the same right-determinacy as the power series without the constant term.
+      ## Therefore remove constant term.
+      R = base_ring(parent(f))
+      shift, _ = base_ring_shifts(parent(f))
+      a = shift(numerator(f))
+      b = shift(denominator(f)) 
+      L, _ = localization(R, complement_of_point_ideal(R, [coefficient_ring(R)(0) for i = 1:ngens(R)]))
+      f_shifted = L(a//b) 
+      f_new = f_shifted-(constant_coefficient(a)//constant_coefficient(b))
+      return determinacy_bound(f_new, equivalence)
+    end
+    X = HypersurfaceGerm(quo(parent(f), ideal(parent(f), f))[1])
+    k = milnor_number(X)
+  else  ## equivalence == :contact
+    ord_f != 0 || return 0
+    k = tjurina_number(f)
+  end
+  return characteristic(parent(f)) == 0 ? k + 1 : 2*k - ord_f + 2
+end
+
+
+
+@doc raw"""
+    determinacy_bound(X::HypersurfaceGerm, equivalence::Symbol = :contact)
+
+Compute some determinacy bound of the hypersurface germ 'X' with respect to ':right' or ':contact' equivalence.
+Return infinity if not finitely determined. 
```

Comment:
Please state explicitly somewhere that PosInf() does not prove the absence of finite determinacy in prime characteristic. In positive characteristic infinity is also returned, if no determinacy bound can be obtained by this approach.

Reply:
In my understanding an upper determinancy bound (strictly less than infinity) can always be computed this way in prime characteristic acording to the paper ''Invariants of Hypersurface Singularities in Positive Characteristic'' by Boubakri, Greuel, Markwig (arXiv:1005.4503). 
The upper bounds provided by Theorem 2.1 respectivly Corollary 2.4 are computated here and as a consequence of Theorem 2.5 these bounds are always finite in the finite determined case. See also Theorem 2.8. Therefore PosInf() is only returned in the non finite determined case.

Reply:
Yes indeed. I take my previous comment back (I had some other invariant in mind).